### ISSUE TITLE
Re-mounting a Field component erases the field-level validation function

### ISSUE THREAD

**Issue by sandkampderek** [erikras/redux-form#3566]
### Are you submitting a **bug report** or a **feature request**?

Bug report

### What is the current behavior?

When a Field with a field-level validation function attached to it is unmounted and immediately re-mounted, that validation function will never be run again.

In the example below, when the key changes, the Field will be unmounted and re-mounted, and the validation function will never be run again:

          <Field
            key={`${this.state.firstNameFieldKey}`}
            component={renderField}
            name="firstName"
            validate={this.validateFirstNameField}
          />

Perhaps a more common method of unmounting and remounting fields is conditional rendering of fields with the same name. In the example below, when this.state.shouldRenderTextArea changes, the validate function will never be run again:

        {
          this.state.shouldRenderTextArea ?
            <Field 
              name="firstName"
              component={renderTextArea}
              validate={this.validateFirstNameField}
            />
            :
            <Field
              name="firstName"
              component={renderTextInput}
              validate={this.validateFirstNameField}
            />
        }

### What is the expected behavior?
After the field is re-mounted, its validation function should run.

### Sandbox Link

https://codesandbox.io/s/m7l1lyzyoy

### What's your environment?

Reproduced in redux-form 7.1.1

### Other information

The problem appears to be that Field registers the validate function in componentWillMount, and erases the validate function in componentWillUnmount. However, React 16 will always run componentWillMount of the new component BEFORE componentWillUnmount of the old component. That means that we register the new validate function, and immediately delete it.

Would there be any problem with having Field register itself it componentDidMount instead?

**Comment by marcneander** [comment]
Status on this?

**Comment by marcneander** [comment]
To add some context: I get the same problem on all 7.x.x versions. Works as expected on 6.8.0

React: 15.6.2

**Comment by Luchanso** [comment]
Yes, I have the same problem. It was because componentWillUnmount call after componentWillMount.

-> componentWillMount First
--- Start Remounting
-> componentWillMount Second
-> componentWillUnmount First

I think redux-form must call register function when componentDidMount. I can try make PR for this?

related links: 
* [How to work validate, register, unregister](https://github.com/erikras/redux-form/blob/master/src/createReduxForm.js#L563-L593)
* [When call register and unregister](https://github.com/erikras/redux-form/blob/master/src/createField.js#L40-L73)

**Comment by Luchanso** [comment]
I resolve it via [destroyOnUnmount](https://redux-form.com/7.2.0/docs/api/reduxform.md/#-code-destroyonunmount-boolean-code-optional-): false, but I don't know what's wrong if you use it in your own project.

**Comment by Guymestef** [comment]
I used destroyOnUnmount: false too as a temporary workaround, but a side-effect is that the form keep the submitted values as initialValues...

**Comment by SengitU** [comment]
Hi folks, I also have this issue. Thanks to @Luchanso , I am using workaround. Do you have any news on this issue?

**Comment by SengitU** [comment]
@Elliot128 we also had same issue. We first thought about clearing values by firing change action, but this approach will keep the validation errors. So we decided to register/deregister fields manually, until we find a good solution.

**Comment by slavab89** [comment]
I believe that https://github.com/erikras/redux-form/issues/3500 also relates to this

**Comment by asaarnak** [comment]
Seems that we should use `componentDidMount`
When using React strict mode:
```
<React.StrictMode>
 <App/>
</React.StrictMode>
```
Then error is printed to console:
`componentWillMount: Please update the following components to use componentDidMount instead: `

**Comment by wanaright10** [comment]
I met this issue as well, but I found a intreating solution:
`
not working below:
const ConditionalComponent = () => someCondition? <ReduxFormFieldComponentOne /> : <ReduxFormFieldComponentTwo />;

render() {
<ConditionalComponent />
}

working below:
not working below:
const ConditionalComponent = () => someCondition? <ReduxFormFieldComponentOne /> : <ReduxFormFieldComponentTwo />;

render() {
{ConditionalComponent()}
}
`
Someone can explain this? The only diff is one use < />, another one use {func()}

**Comment by asaarnak** [comment]
As redux-form still uses `unsafeComponentWillMount`. See https://github.com/erikras/redux-form/blob/master/src/createField.js#L41
I think this isn't fixed.
This issue initial comment. See: https://github.com/erikras/redux-form/issues/3566#issue-268524710
```
The problem appears to be that Field registers the validate function in componentWillMount, and erases the validate function in componentWillUnmount. However, React 16 will always run componentWillMount of the new component BEFORE componentWillUnmount of the old component. That means that we register the new validate function, and immediately delete it.

Would there be any problem with having Field register itself it componentDidMount instead?
```

**Comment by kumarsaranya** [comment]
Hi,

is this issue resolved? I am facing similar issue and still need to depend on onDestroyUnmount: false workaround.

**Comment by F-Adam-B** [comment]
Hi,
Having sort of the same issue here. Using version 7.4.2 and when setting field level validation it not only doesn't error, the entire field disappears if I don't input a value and tab through the the next field. onDestroyUnmount isn't helping either.

thanks for any help to resolve these issues

**Comment by sergeyromanovsky** [comment]
This issue is stil reproducable on redux-form 8.2
